File: runtime/cs/bintalk/ProtocolWriter.cs
using System;
using System.Text;
using System.Collections.Generic;

namespace bintalk
{
    /** This class can write basic types by using a bintalk.IWriter object. */
    public static partial class ProtocolWriter
    {
        public static void writeMid(bintalk.IWriter w, int v)
        {
            write(w, (ushort)v);
        }
        // int 64
    	public static void write(bintalk.IWriter w, long v)   
        { 
            w.write(BitConverter.GetBytes(v)); 
        }
        // uint64
    	public static void write(bintalk.IWriter w, ulong v)  
        { 
            w.write(BitConverter.GetBytes(v)); 
        }
        // double
    	public static void write(bintalk.IWriter w, double v) 
        { 
            w.write(BitConverter.GetBytes(v)); 
        }
        // float
    	public static void write(bintalk.IWriter w, float v)  
        { 
            w.write(BitConverter.GetBytes(v)); 
        }
        // int32
    	public static void write(bintalk.IWriter w, int v)    
        { 
            w.write(BitConverter.GetBytes(v));
        }
        // uint32
    	public static void write(bintalk.IWriter w, uint v)   
        {
            w.write(BitConverter.GetBytes(v));
        }
        // int16
    	public static void write(bintalk.IWriter w, short v)  
        { 
            w.write(BitConverter.GetBytes(v));
        }
        // uint16
    	public static void write(bintalk.IWriter w, ushort v) 
        {
            w.write(BitConverter.GetBytes(v));
        }
        // int8
    	public static void write(bintalk.IWriter w, sbyte v)  
        {
            w.write(new byte[1]{(byte)v});
        }
        // uint8
    	public static void write(bintalk.IWriter w, byte v)   
        {
            w.write(new byte[1]{v});
        }
        // bool
    	public static void write(bintalk.IWriter w, bool v)   
        {
            write(w, (byte)(v?1:0));
        }
        // string
    	public static void write(bintalk.IWriter w, string v)
    	{
            if (v == null || v.Length == 0)
                writeDynSize(w, 0);
            else
            {
                byte[] str = Encoding.UTF8.GetBytes(v);
                uint len = (uint)str.Length;
                writeDynSize(w, len);
                if (len > 0)
                    w.write(str);
            }
    	}
        // binary.
        public static void write(bintalk.IWriter w, byte[] v) 
        { 
            if (v == null || v.Length == 0)
                writeDynSize(w, 0);
            else
            {
                uint s = (uint)v.Length;
                writeDynSize(w, s);
                w.write(v); 
            }
        }
        // dynamic size.
        public static void writeDynSize(bintalk.IWriter w, uint s)
        {
            byte[] b = BitConverter.GetBytes(s);
            int n = 0;
            if (s <= 0X3F)
                n = 0;
            else if (s <= 0X3FFF)
                n = 1;
    		else if(s <= 0X3FFFFF)
    			n = 2;
    		else if(s <= 0X3FFFFFFF)
    			n = 3;
            b[n] |= (byte)(n << 6);
    		for(int i = n; i >= 0; i--)
    			write(w, b[i]);
        }
    }
}

File: runtime/cpp/bintalk/ProtocolWriter.cpp
#include "Config.h"
#include "ProtocolWriter.h"

namespace bintalk
{

void ProtocolWriter::writeMID(BinaryWriter* w, MID mid)
{
#if BINTALK_BIG_ENDIAN
	mid = swapEndian<MID>(mid);
#endif//BINTALK_BIG_ENDIAN
	w->write(&mid, sizeof(MID));
}

#if BINTALK_BIG_ENDIAN
	#define BINTALK_PROTOCOLWRITER_SINGLE_IMP(T)\
		void ProtocolWriter::write##T(BinaryWriter* w, const T& v) \
		{ T vv = swapEndian<T>(v); w->write(&vv, sizeof(T)); }
#else //BINTALK_BIG_ENDIAN 
	#define BINTALK_PROTOCOLWRITER_SINGLE_IMP(T)\
		void ProtocolWriter::write##T(BinaryWriter* w, const T& v) \
		{ w->write(&v, sizeof(T)); }
#endif //BINTALK_BIG_ENDIAN 

#define BINTALK_PROTOCOLWRITER_ARRAY_IMP(T)\
void ProtocolWriter::write##T##A(BinaryWriter* w, const std::vector<T>& v)\
{\
	UINT32 s = (UINT32)v.size();\
	writeDynSize(w, s);\
	for(UINT32 i = 0; i < s; i++)\
		write##T(w, v[i]);\
}
#define BINTALK_PROTOCOLWRITER_SIMPLE_IMP(T)\
	BINTALK_PROTOCOLWRITER_SINGLE_IMP(T);\
	BINTALK_PROTOCOLWRITER_ARRAY_IMP(T);

BINTALK_PROTOCOLWRITER_SIMPLE_IMP(INT64);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(UINT64);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(DOUBLE);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(FLOAT);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(INT32);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(UINT32);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(INT16);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(UINT16);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(INT8);
BINTALK_PROTOCOLWRITER_SIMPLE_IMP(UINT8);

void ProtocolWriter::writeBOOL(BinaryWriter* w, const BOOL& v)
{
	UINT8 v1 = v?1:0;
	w->write(&v1, sizeof(UINT8));
}
BINTALK_PROTOCOLWRITER_ARRAY_IMP(BOOL);
void ProtocolWriter::writeSTRING(BinaryWriter* w, const STRING& v)
{
	UINT32 s = (UINT32)v.size();
	writeDynSize(w, s);
	w->write(v.c_str(), s);
}
BINTALK_PROTOCOLWRITER_ARRAY_IMP(STRING);
void ProtocolWriter::writeBINARY(BinaryWriter* w, const BINARY& v)
{
	UINT32 s = (UINT32)v.size();
	writeDynSize(w, s);
	if(s > 0)
		w->write(&(v[0]), s);
}
BINTALK_PROTOCOLWRITER_ARRAY_IMP(BINARY);
// Dynamic size.
void ProtocolWriter::writeDynSize(BinaryWriter* w, UINT32 s)
{
	UINT8* p = (UINT8*)(&s);
	UINT8 n = 0;
	if(s <= 0X3F)
		n = 0;
	else if(s <= 0X3FFF)
		n = 1;
	else if(s <= 0X3FFFFF)
		n = 2;
	else if(s <= 0X3FFFFFFF)
		n = 3;
	p[n] |= (n<<6);
	for(int i = (int)n; i >= 0; i--)
		w->write(p+i, sizeof(UINT8));
}

}